File: src/chapters/introduction/content/code/horner.go
package main

import "fmt"

// Evaluate a polynomial with coefficients a = [a_n, ..., a_0] (high -> low) at x.
func horner(a []float64, x float64) float64 {
	y := a[0]
	for _, c := range a[1:] {
		y = y*x + c
	}
	return y
}

func main() {
	// p(x) = 5x^4 - 8x^3 + 2x^2 + 4x - 10
	fmt.Println(horner([]float64{5, -8, 2, 4, -10}, 2)) // 22
}

File: src/chapters/introduction/content/code/horner.cpp
#include <vector>
#include <iostream>
using namespace std;

// Evaluate a polynomial with coefficients a = {a_n, ..., a_0} (high -> low) at x.
double horner(const vector<double>& a, double x) {
    double y = a[0];
    for (size_t i = 1; i < a.size(); ++i) y = y * x + a[i];
    return y;
}

int main() {
    cout << horner({5, -8, 2, 4, -10}, 2) << "\n";   // 22
}
